File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    answer = 0
    n=len(triangle)
    dp=[[0]*i for i in range(1,n+1)]
    dp[0][0]=triangle[0][0]
    
    for i in range(1,n):
        size=len(triangle[i])
        
        for j in range(i+1):
            if j==0:
                dp[i][j]=dp[i-1][j]+triangle[i][j]
            elif j==(size-1):
                dp[i][j]=dp[i-1][j-1]+triangle[i][j]
            else:
                dp[i][j]= max(dp[i-1][j-1],dp[i-1][j]) + triangle[i][j]
        
    return max(dp[-1])

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;
class Solution {
    public int solution(int[][] triangle) {
        for (int i=1; i<triangle.length; i++){
            triangle[i][0]+=triangle[i-1][0];
            triangle[i][i]+=triangle[i-1][i-1];
            for (int j=1; j<triangle[i].length-1; j++){
                triangle[i][j]+=Math.max(triangle[i-1][j-1],triangle[i-1][j]);
            }
        }
        
        int answer = 0;
        int n = triangle.length;
        for (int i=0; i<triangle[n-1].length;i++){
            answer = Math.max(triangle[n-1][i],answer);
        }
        return answer;
    }
}
